### PR TITLE
Migrate VPA metrics to CustomResourceState metrics and upgrade `kube-state-metrics` to `v2.13.0` 

### PR DESCRIPTION
**Pull request by vicwicker**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:

This PR allows the upgrade of the kube-state-metrics to the latest version, as of today v2.13.0. The upgrade was not possible because of the deprecation of the VPA metrics, which are now supposed to be configured as any other custom resource by using a so-called [`CustomResourceState`](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md) feature from the kube-state-metrics. This is a mechanism to tell kube-state-metrics via configuration which keys and values from custom resource spec files need to be mapped into which metrics. 

We tried to preserve the old metric labels and naming as much as possible. However, there are few changes. The most significant is that the resource (i.e., CPU or memory) is not a label anymore in the new metrics but part of the metric name.

<details><summary>Example for the blackbox-exporter CPU target recommendation</summary>

```
{
    "metric": {
      "__name__": "kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target",
      "container": "blackbox-exporter",
      "instance": "kube-state-metrics",
      "job": "kube-state-metrics",
      "namespace": "kube-system",
      "resource": "cpu",
      "target_api_version": "apps/v1",
      "target_kind": "Deployment",
      "target_name": "blackbox-exporter",
      "type": "shoot",
      "unit": "core",
      "verticalpodautoscaler": "blackbox-exporter"
    }
},
```

becomes

```
{
  "__name__": "kube_customresource_verticalpodautoscaler_status_recommendation_containerrecommendations_target_cpu",
  "container": "blackbox-exporter",
  "customresource_group": "autoscaling.k8s.io",
  "customresource_kind": "VerticalPodAutoscaler",
  "customresource_version": "v1",
  "instance": "kube-state-metrics",
  "job": "kube-state-metrics",
  "namespace": "kube-system",
  "target_api_version": "apps/v1",
  "target_kind": "Deployment",
  "target_name": "blackbox-exporter",
  "type": "shoot",
  "unit": "core",
  "verticalpodautoscaler": "blackbox-exporter"
}
```

</details>

This PR also improves the deployment of kube-state-metrics. In short, the most relevant enhancements are:

1. Move away from the concept of `ResourceConfigs` and adopt `ManagedResource` to deploy all kube-state-metrics artifacts. Code-wise this resembles now the way Prometheus is deployed.
2. Use the kube-state-metrics filtering capabilities to make the Prometheus scrape jobs lighter. Currently, unwanted metrics are also scraped but discarded during the scrape job relabeling step.

As for the unit tests, the `CustomResourceState` expected configuration is stored in a YAML file and committed to the repository. The `expectedCustomResourceStateConfig()` helper function is to be used when loading such expectation from disk during test execution. It loads this file as well as it generates the configuration the code being tested produces. This actual configuration is stored in a file in the temporary directory and matched against the expectation. If the match succeeds, the test will run. If not, the test will be aborted and a help message will be printed by ginkgo with a `diff` command to compare both files. Developers can compare both files with any tool of their choice and fix the test accordingly. Reviewers can find below and example of this test output.

<details>
<summary>Example of test failure</summary>

Screenshot 2024-06-06 at 16 30 03

</details>

**Special notes for your reviewer**:

/cc @istvanballok @rickardsjp 
/fyi @rfranzke @voelzmo 

**Release note**:

```other operator
Migrate VPA metrics to CustomResourceState metrics and upgrade `kube-state-metrics` to `v2.13.0`.
```

### CODE REVIEW

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go [hunk 1]
```diff
@@ -37,6 +37,156 @@ import (
 	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
 )
 
+/*
+TODO(vicwicker): The documentation instructions does not include the kind but the examples do. A quick test
+showed the kind is superfluous when --custom-resource-state-config is given. We need a bit more investigation
+in case there is a resource of kind CustomResourceStateMetrics that we could use to pass this lenghty
+configuration more elegantly.
+*/
```
Comment (rickardsjp) [inline pkg/component/observability/monitoring/kubestatemetrics/resources.go]: Have you had a look [at the API](https://pkg.go.dev/k8s.io/kube-state-metrics/v2@v2.12.0/pkg/customresourcestate)? Looks like it should be possible to avoid string literals and do it in code :)

Comment (vicwicker) [reply pkg/component/observability/monitoring/kubestatemetrics/resources.go]: Yes. I moved to generate this configuration programatically some commits later :)

Comment (rickardsjp) [reply pkg/component/observability/monitoring/kubestatemetrics/resources.go]: Oops, that's what I get for going through the commits but stopping short of the end.

**Comment by vicwicker** [comment]
/unhold
/retest-required

This is now ready for review.

/cc @chrkl @rickardsjp @istvanballok

File: go.mod [hunk 2]
```diff
@@ -74,17 +74,19 @@ require (
...
 	k8s.io/pod-security-admission v0.29.7
 	k8s.io/utils v0.0.0-20240711033017-18e509b52bc8
 	sigs.k8s.io/controller-runtime v0.17.5
-	sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20231015215740-bf15e44028f9 // v0.16.3
 	sigs.k8s.io/controller-tools v0.14.0
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1
 	sigs.k8s.io/yaml v1.4.0
 )
 
+require sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20231015215740-bf15e44028f9 // v0.16.3
```
Comment (rfranzke) [inline go.mod]: Why did you move this out of the above `require` block? Please move it back.

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go [hunk 3]
```diff
@@ -107,30 +156,77 @@ func (k *kubeStateMetrics) Deploy(ctx context.Context) error {
...
+	}
+
+	if k.values.ClusterType == component.ClusterTypeShoot {
+		registryTarget := managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
+		resourcesTarget, err := registryTarget.AddAllAndSerialize(k.getResourcesForShootTarget(shootAccessSecret)...)
+		if err != nil {
+			return err
+		}
+
+		// TODO(vicwicker): Remove after KSM upgrade to 2.13.
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]:
```suggestion
		// TODO(vicwicker): Remove after Gardener v1.104 got released.
```

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go [hunk 4]
```diff
@@ -736,3 +712,14 @@ func (k *kubeStateMetrics) nameSuffix() string {
 	}
 	return suffix + k.values.NameSuffix
 }
+
+func (k *kubeStateMetrics) customResourceStateConfigMap() *corev1.ConfigMap {
+	customResourceStateConfig, err := yaml.Marshal(NewCustomResourceStateConfig())
+	utilruntime.Must(err)
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/resources.go]: We don't want to panic, please return the error instead.

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go [hunk 5]
```diff
@@ -736,3 +712,14 @@ func (k *kubeStateMetrics) nameSuffix() string {
...
+	cm := &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: customResourceStateConfigMapNamePrefix, Namespace: k.namespace}}
+	cm.Data = map[string]string{
+		customResourceStateConfigMountFile: string(customResourceStateConfig),
+	}
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/resources.go]:
```suggestion
	configMap := &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{Name: customResourceStateConfigMapNamePrefix, Namespace: k.namespace},
		Data: map[string]string{
			customResourceStateConfigMountFile: string(customResourceStateConfig),
		},
	}
```

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go [hunk 6]
```diff
@@ -77,21 +76,71 @@ type Values struct {
...
+	if k.values.NameSuffix == SuffixSeed {
+		scrapeConfig = k.scrapeConfigSeed()
+	} else if k.values.NameSuffix == SuffixRuntime {
+		scrapeConfig = k.scrapeConfigGarden()
+	}
+
+	clusterRole := k.clusterRole()
+	serviceAccount := k.serviceAccount()
+	customResourceStateConfigMap := k.customResourceStateConfigMap()
+	deployment := k.deployment(serviceAccount, "", nil, customResourceStateConfigMap.Name)
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]: var () block

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go [hunk 7]
```diff
@@ -77,21 +76,71 @@ type Values struct {
...
+		scrapeConfig,
+		k.service(),
+		k.verticalPodAutoscaler(deployment),
+		customResourceStateConfigMap,
+	}
+}
+
+func (k *kubeStateMetrics) getResourcesForShoot(genericTokenKubeconfigSecretName string, shootAccessSecret *gardenerutils.AccessSecret) []client.Object {
+	customResourceStateConfigMap := k.customResourceStateConfigMap()
+	deployment := k.deployment(nil, genericTokenKubeconfigSecretName, shootAccessSecret, customResourceStateConfigMap.Name)
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]: var () block

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go (hunk 3 above)
Comment (rfranzke) [reply pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]: If we want to safely remove this code in 3 releases, I guess we should add some extra deploy step for `kube-state-metrics` after https://github.com/gardener/gardener/blob/1832cc548ca2cf8eb9bd2b2454728f217ec14f01/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L132-L135 in case `kube-state-metrics` already exists in the shoot namespace.
This way, we can be sure that this code has been called for all shoots.

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go [hunk 8]
```diff
@@ -544,168 +582,104 @@ func (k *kubeStateMetrics) reconcileScrapeConfigGarden(scrapeConfig *monitoringv
...
+		{
+			Record: "shoot:machine_types:sum",
+			Expr:   intstr.FromString(`sum(kube_node_labels) by (label_beta_kubernetes_io_instance_type)`),
+		},
+		{
+			Record: "shoot:node_operating_system:sum",
+			Expr:   intstr.FromString(`sum(kube_node_info) by (os_image, kernel_version)`),
+		},
+		// Mitigation for extension dashboards.
+		// TODO(istvanballok): Remove in a future version. For more details, see https://github.com/gardener/gardener/pull/6224.
```
Comment (rfranzke) [inline pkg/component/observability/monitoring/kubestatemetrics/resources.go:666]: When can this be removed exactly?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice PR, thanks for the refactoring to GRM-only management :)

File: go.mod (hunk 2 above)
Comment (vicwicker) [reply go.mod]: I tried to capture the reasons in commit [`d281b7b`](https://github.com/gardener/gardener/pull/9941/commits/d281b7ba2f0a7efad3065c98d464b13126d3b92b). 

But in short, when this line is inside the `require` block from above, go is not able to understand that such block is for the direct dependencies and that new direct dependencies should be added there as well. So, when I added the KSM dependency via `go get` and `go mod tidy`,  go created a new `require` block only for the KSM. @istvanballok and I investigated this and found the culprit is this dependency because it has the version commented at the end (`... // v0.16.3`). After extracting this dependency in a new `require` block, `go` correctly added the new KSM dependency in the right block.

But this was just a proposal. We can keep it in one block altogether. I will address this 👍🏻

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go (hunk 4 above)
Comment (vicwicker) [reply pkg/component/observability/monitoring/kubestatemetrics/resources.go]: Yes, you're right. I didn't return the error in the first place because the configuration that could error here is statically generated. But it definitely feels better to return the error and be on the safe-side 👍🏻

File: pkg/component/observability/monitoring/kubestatemetrics/resources.go (hunk 8 above)
Comment (vicwicker) [reply pkg/component/observability/monitoring/kubestatemetrics/resources.go:666]: Unfortunately I can't answer that myself. Maybe we could ping @istvanballok and trigger the discussion.

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go (hunk 7 above)
Comment (vicwicker) [reply pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]: After addressing [this comment](https://github.com/gardener/gardener/pull/9941#discussion_r1713636431) the code now looks like the following. Unfortunately, I am not sure if a var block is still meaningful after the change. I would rather leave it like this if that's okay with you.

```
func (k *kubeStateMetrics) getResourcesForShoot(genericTokenKubeconfigSecretName string, shootAccessSecret *gardenerutils.AccessSecret) ([]client.Object, error) {
	customResourceStateConfigMap, err := k.customResourceStateConfigMap()
	if err != nil {
		return nil, err
	}

	deployment := k.deployment(nil, genericTokenKubeconfigSecretName, shootAccessSecret, customResourceStateConfigMap.Name)
	return []client.Object{
		...
	}, nil
}
```

File: go.mod (hunk 2 above)
Comment (rfranzke) [reply go.mod]: Does it work with just one block if the comment is not at the end but in the line before?

Comment (vicwicker) [reply go.mod]: I just tried. Unfortunately it does not :(

Comment (rfranzke) [reply go.mod]: OK. I can live with both, either put it back, or keep it in the other block, but then maybe add a short comment why this is (to prevent we move this back up later)

Comment (vicwicker) [reply go.mod]: I put it back. I think it's going to be more confusing than helpful :)

**Comment by rickardsjp** [comment]
> LGTM label has been added.

just for posterity's sake: this was from me, I had missed the comments just before requesting changes. I thought I was quick enough to beat the bot but I was not. I'll re-add LGTM when you push the changes, @vicwicker :)

File: pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go (hunk 3 above)
Comment (vicwicker) [reply pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go]: This should be addressed now.

**Comment by vicwicker** [comment]
This is again ready for review.

/unhold

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go [hunk 9]
```diff
@@ -133,6 +133,13 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 			Name: "Deploying Shoot namespace in Seed",
 			Fn:   flow.TaskFn(botanist.DeploySeedNamespace).RetryUntilTimeout(defaultInterval, defaultTimeout),
 		})
+		// TODO(vicwicker): Remove after Gardener v1.104 got released.
+		_ = g.Add(flow.Task{
+			Name:         "Reconciling kube-state-metrics for Shoot in Seed for the monitoring stack during migration to Gardener v1.104",
+			Fn:           flow.TaskFn(botanist.DeployKubeStateMetrics).RetryUntilTimeout(defaultInterval, 2*time.Minute),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go]:
```suggestion
			Fn:           flow.TaskFn(func(ctx context.Context) error {
				// check here if the `kube-state-metrics` deployment exists in the shoot namespace. if not, return nil
				return botanist.DeployKubeStateMetrics(ctx)
			}).RetryUntilTimeout(defaultInterval, 2*time.Minute),
```

We don't want to prematurely deploy KSM into the shoot namespace if it does not exist yet. We just want to migrate it early if it already exists. Please adapt :)

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go (hunk 9 above)
Comment (vicwicker) [reply pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go]: It should be addressed now.

**Review by rfranzke — COMMENTED** [review]
Thank you @vicwicker !

/lgtm
/approve

**Review by rickardsjp — APPROVED** [review]
/lgtm
